### DISCUSSION TITLE
Connect by bluetooth or USB serial instead of wifi

### DISCUSSION THREAD

**Discussion by janifr** [SignalK/SensESP#680] [General]
Hi,

I have Signal K installed on Raspberry PI 4 and would prefer to connect ESP32 board by bluetooth or wired USB serial connection instead of wifi, because I want to use ADC2 channels, which don't work with wifi enabled.

Is there a way to do it?

TIA, Jani

**Comment by mairas** [comment]
There's a bit of extra coding to be done, but it should be possible. In the WiFi case, you have the following players:

- `Networking` class in `sensesp/net/networking.h` takes care of the WiFi connectivity
- `WSClient` class in `sensesp/net/ws_client.h` manages the SK websocket connection (and the related auth requests).
- `SKDeltaQueue` in `sensesp/signalk/signalk_delta_queue.h` manages the outbound message queue.
- `SKEmitter` (and `SKOutput`) in `sensesp/signalk/signalk_emitter.h` (`signalk_output.h`) serializes data into SK Deltas.

There can be multiple `SKDeltaQueue` objects (one for each connection method). The queues observe all emitters and stores the deltas in a queue.

`WSClient` polls the delta queue and if there are any deltas available, sends them to the server.

If you want to do this without WiFi, instantiate your app using `SensESPMinimalApp` to avoid the `Networking` and `WSCLient` objects. Then, you need to instantiate a `SKDeltaQueue` object manually and poll it repeatedly and send outbound deltas either over the default `Serial` object or your own Bluetooth serial connection.

Also, for USB serial, you need to disable serial debug logging (or move it over to `Serial1`).

If you decide to do this, having some output in a form of PR (example code, additional functionality, documentation) would be absolutely great!

**Comment by ba58smith** [comment]
@janifr , even if you switch to Bluetooth or serial, I think you'll still need to incorporate this PR in order to use any of the ADC2 pins: https://github.com/SignalK/SensESP/pull/642
